data_model/derive/src/id.rs

```diff
@@ -0,0 +1,120 @@
+#![allow(clippy::expect_used, clippy::eval_order_dependence)]
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream},
+    Attribute, Field, Generics, Ident, Token, TypePath, Visibility,
+};
+
+pub(super) struct IdInput {
+    ident: Ident,
+    id_type: TypePath,
+}
+
+impl Parse for IdInput {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let attrs = input.call(Attribute::parse_outer)?;
+        let id_type = parse_id_attribute(&attrs);
+
+        let _vis = input.parse::<Visibility>()?;
+        let _struct_token = input.parse::<Token![struct]>()?;
+        let ident = input.parse()?;
+        let _generics = input.parse::<Generics>()?;
+        let content;
+        let _brace_token = syn::braced!(content in input);
+        let _struct_fields = content.parse_terminated::<Field, Token![,]>(Field::parse_named)?;
+
+        Ok(IdInput { ident, id_type })
+    }
+}
+
+fn impl_ordeqhash(ast: &IdInput) -> proc_macro2::TokenStream {
+    let name = ast.ident.clone();
+
+    quote! {
+        impl core::cmp::PartialOrd for #name {
+            #[inline]
+            fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+                Some(self.cmp(other))
+            }
+        }
+
+        impl core::cmp::Ord for #name {
+            fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+                self.id().cmp(other.id())
+            }
+        }
+
+        impl core::cmp::PartialEq for #name {
+            fn eq(&self, other: &Self) -> bool {
+                self.id() == other.id()
+            }
+        }
+
+        impl core::cmp::Eq for #name {}
+
+        impl core::hash::Hash for #name {
+            fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+                self.id().hash(state);
+            }
+        }
+    }
+}
+
+pub(super) fn impl_id(ast: &IdInput) -> TokenStream {
+    let id = ast.id_type.clone();
+    let name = ast.ident.clone();
+
+    let ordeqhash = impl_ordeqhash(ast);
+
+    let body = if ast.ident.to_string().starts_with("NewRole") {
```

Comment:
uh, I'm not happy about this hardcoding. Can't we do something about it?

Reply:
We can also introduce macro attribute so that it will look like:

```rust
#[derive(IdOrdEqHash)]
#[id(type = "Id")]
#[id(id = "inner.id")]
pub struct NewRole {
    inner: Role,
}
```

Reply:
This can be another issue btw

Reply:
Created #2522.